File: packages/web/tests/sharing-seed.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import { GoogleDriveRemote } from "../src/vault/drive";
import { createRecoveryCipher } from "../src/vault/crypto";
import { parseRecoveryCode } from "../src/vault/recovery";
import { FakeDrive } from "./helpers";

/**
 * The seed that makes one person one participant, read across platforms.
 *
 * Sharing pins one key per person. A keyring shared from a phone has to be
 * manageable from that person's laptop, and the people they shared with
 * trusted one key — so every device of theirs has to derive the same sharing
 * identity, which means arriving at the same seed.
 *
 * Until now the only thing carrying that seed was the printed recovery code,
 * and nothing distributed it. A browser holding the passkey could read every
 * password in the vault and still not touch a shared keyring, because it had
 * never been handed a piece of paper.
 *
 * `fixtures/drive-with-sharing-seed-v1.json` is written by Android's real
 * `DriveVaultRemote`. This opens it with the browser's real reader.
 */
const fixture = JSON.parse(
  readFileSync(
    new URL("../../../fixtures/drive-with-sharing-seed-v1.json", import.meta.url),
    "utf8",
  ),
) as { recoveryCode: string; content: string };

function driveHoldingIt() {
  const drive = new FakeDrive();
  drive.files.set("file-1", {
    name: "keyweb-vault-v1.json",
    content: fixture.content,
    revision: 1,
    appProperties: { keyweb: "vault-v1" },
  });
  return drive;
}

describe("a sharing seed a phone published", () => {
  it("is not in the file in the clear", () => {
    expect(fixture.content).not.toContain(fixture.recoveryCode);
    expect(fixture.content).not.toContain(fixture.recoveryCode.replace(/-/g, ""));
  });

  it("opens in a browser holding the same code", async () => {
    const drive = driveHoldingIt();
    const secret = parseRecoveryCode(fixture.recoveryCode);
    const payload = JSON.parse(fixture.content) as { seed: { recovery: unknown } };
    const recoveryCipher = await createRecoveryCipher(secret, payload.seed.recovery, "localhost");

    const remote = new GoogleDriveRemote({
      clientId: "test",
      cipher: recoveryCipher,
      recoveryCipher,
      store: drive.asStore(),
      authorize: async () => ({ accessToken: "token" }) as never,
    });

    const seed = await remote.sharingSeed();
    expect(seed).not.toBeNull();
    // The same bytes the phone derives its identity from — which is what makes
    // the two devices one participant rather than two.
    expect([...seed!]).toEqual([...secret]);
  });

  it("is absent, not invented, when the file has none", async () => {
    const drive = new FakeDrive();
    drive.files.set("file-1", {
      name: "keyweb-vault-v1.json",
      content: JSON.stringify({ v: 1, recovery: JSON.parse(fixture.content).recovery }),
      revision: 1,
      appProperties: { keyweb: "vault-v1" },
    });
    const secret = parseRecoveryCode(fixture.recoveryCode);
    const recoveryCipher = await createRecoveryCipher(
      secret,
      JSON.parse(fixture.content).recovery,
      "localhost",
    );
    const remote = new GoogleDriveRemote({
      clientId: "test",
      cipher: recoveryCipher,
      recoveryCipher,
      store: drive.asStore(),
      authorize: async () => ({ accessToken: "token" }) as never,
    });

    expect(await remote.sharingSeed()).toBeNull();
  });
});

File: packages/android/app/src/test/java/app/keyweb/data/SharingSeedTest.kt
package app.keyweb.data

import app.keyweb.vault.RecoveryCode
import app.keyweb.vault.VaultEnvelopeCipher
import app.keyweb.vault.VaultOp
import app.keyweb.vault.applyOp
import app.keyweb.vault.emptyVault
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject

/**
 * The seed that makes one person one participant on all their devices.
 *
 * Sharing pins one key per person: the people you shared with trust that key,
 * and a keyring shared from your phone has to be manageable from your laptop.
 * The identity is wrapped with a key derived from this seed, so every device
 * has to arrive at the same bytes — and until now the only thing that carried
 * them was the printed recovery code, which nothing distributed. A browser
 * that had never been handed that code could read every password in the vault
 * and still not touch a shared keyring.
 *
 * So it rides in the backup, sealed under both keys, and a device takes it
 * from whichever copy it can open.
 */
class SharingSeedTest {

    private val json = Json { ignoreUnknownKeys = true }

    private fun vault() = applyOp(
        emptyVault(),
        VaultOp.KeyringPut("k", "001700000000000-00000-t", "ring", "Household"),
    )

    @Test
    fun `is written into the backup, sealed so either key opens it`() = runTest {
        val drive = FakeDrive()
        val secret = RecoveryCode.generate()
        val remote = DriveVaultRemote(
            drive,
            VaultEnvelopeCipher.forRecoveryCode(secret),
            sharingSeed = secret,
        )
        remote.write(vault(), null)

        val payload = json.parseToJsonElement(assertNotNull(drive.vaultFile()).content).jsonObject
        val seed = assertNotNull(payload["seed"]).jsonObject
        assertTrue(seed.containsKey("recovery"))
        // Nothing about it is in the clear.
        assertTrue(!assertNotNull(drive.vaultFile()).content.contains(RecoveryCode.format(secret)))
    }

    @Test
    fun `comes back out through the same code that sealed it`() = runTest {
        val drive = FakeDrive()
        val secret = RecoveryCode.generate()
        DriveVaultRemote(
            drive,
            VaultEnvelopeCipher.forRecoveryCode(secret),
            sharingSeed = secret,
        ).write(vault(), null)

        val payload = json.parseToJsonElement(assertNotNull(drive.vaultFile()).content).jsonObject
        val sealed = json.decodeFromJsonElement(
            app.keyweb.vault.SyncEnvelopeV1.serializer(),
            assertNotNull(payload["seed"]).jsonObject.getValue("recovery"),
        )
        val cipher = VaultEnvelopeCipher.forRecoveryCode(secret, sealed)
        assertContentEquals(secret, cipher.openBytes(sealed))
    }

    @Test
    fun `is never replaced once the file has one`() = runTest {
        val drive = FakeDrive()
        val first = RecoveryCode.generate()
        DriveVaultRemote(
            drive,
            VaultEnvelopeCipher.forRecoveryCode(first),
            sharingSeed = first,
        ).write(vault(), null)
        val before = assertNotNull(drive.vaultFile()).content

        // A device that knows a *different* seed must carry the file's forward:
        // two seeds is two identities, and one person cannot be two participants.
        val second = RecoveryCode.generate()
        val head = DriveVaultRemote(drive, VaultEnvelopeCipher.forRecoveryCode(first))
            .write(vault(), null)
        DriveVaultRemote(
            drive,
            VaultEnvelopeCipher.forRecoveryCode(first),
            sharingSeed = second,
        ).write(vault(), head)

        val after = json.parseToJsonElement(assertNotNull(drive.vaultFile()).content).jsonObject
        val sealed = json.decodeFromJsonElement(
            app.keyweb.vault.SyncEnvelopeV1.serializer(),
            assertNotNull(after["seed"]).jsonObject.getValue("recovery"),
        )
        assertContentEquals(
            first,
            VaultEnvelopeCipher.forRecoveryCode(first, sealed).openBytes(sealed),
        )
        assertTrue(before.contains("\"seed\""))
    }

    @kotlinx.serialization.Serializable
    private data class Fixture(
        val note: String,
        val recoveryCode: String,
        val content: String,
    )

    /**
     * And the browser has to be able to open it, which is the whole point:
     * written here by the shipping writer, read there by the shipping reader.
     */
    @Test
    fun `emits a backup carrying a seed, for the web suite`() = runTest {
        val drive = FakeDrive()
        val secret = RecoveryCode.generate()
        DriveVaultRemote(
            drive,
            VaultEnvelopeCipher.forRecoveryCode(secret),
            sharingSeed = secret,
        ).write(vault(), null)

        val fixtures = java.io.File("../../../fixtures")
        fixtures.mkdirs()
        java.io.File(fixtures, "drive-with-sharing-seed-v1.json").writeText(
            Json { prettyPrint = true }.encodeToString(
                Fixture.serializer(),
                Fixture(
                    note = "Written by Android's DriveVaultRemote. Do not edit.",
                    recoveryCode = RecoveryCode.format(secret),
                    content = assertNotNull(drive.vaultFile()).content,
                ),
            ),
        )
    }

    @Test
    fun `a phone with no seed leaves the file alone`() = runTest {
        val drive = FakeDrive()
        val secret = RecoveryCode.generate()
        DriveVaultRemote(drive, VaultEnvelopeCipher.forRecoveryCode(secret)).write(vault(), null)

        val payload = json.parseToJsonElement(assertNotNull(drive.vaultFile()).content).jsonObject
        assertNull(payload["seed"])
    }
}
